### PR TITLE
Save CARBUS Analyzer format

### PR DESCRIPTION
**Pull request by KostyaSha**
Related https://github.com/collin80/SavvyCAN/pull/243

### CODE REVIEW

**Comment by KostyaSha** [comment]
Rebased locally to latest master, but timestamps are broken, so put on delay this PR

File: framefileio.cpp [hunk 1]
```diff
@@ -817,6 +823,114 @@ bool FrameFileIO::loadCARBUSAnalyzerFile(QString filename, QVector<CANFrame>* fr
...
+
+    return finalStr;
+}
+
+bool FrameFileIO::saveCARBUSAnalzyer(QString filename, const QVector<CANFrame>* frames)
+{
+    QFile *outFile = new QFile(filename);
+    int lineCounter = 0;
+
+    uint64_t minTime = frames->at(0).timeStamp().microSeconds();
```
Comment (KostyaSha) [inline framefileio.cpp:848]: loosing seconds

**Comment by collin80** [comment]
I messed up the time stamp, it's true. It should be something like this:
                double timeStamp = time.toDouble() * 1000000.0;
                thisFrame.setTimeStamp(QCanBusFrame::TimeStamp(0, static_cast<int64_t>(timeStamp)));
Which uses just microseconds for storing the time stamp. For the life of me, I don't know why I did that. I ought to be normalizing the time stamps to seconds and microseconds separately instead of using just microseconds. But, since both are 64 bit integers using just microseconds is still pretty inclusive. 2^64 microseconds is about 1/2 a million years.

**Comment by KostyaSha** [comment]
I think their idea was that microseconds should contain only microseconds, so everything bigger (seconds) is moved to seconds... with such logic they can also introduce hours and re-implement Date...
Imho better make some utils and wrap all this magic into

**Comment by KostyaSha** [comment]
@collin80 could you fixup master so i can continue working on adding and testing .trc format?

**Comment by collin80** [comment]
The fix for the timestamp problem is merged in now. I still haven't normalized the timestamps yet so they all are just in microseconds and nothing is stored in the seconds field. This will be fixed at some point. I'll go add an issue for this right now.

**Comment by KostyaSha** [comment]
found also information about version formats, will fix CARBUS load in scope of this PR soon.

File: framefileio.cpp [hunk 2]
```diff
@@ -751,30 +758,40 @@ bool FrameFileIO::isCARBUSAnalyzerFile(QString filename)
...
+//            sec,us - for version 3
 bool FrameFileIO::loadCARBUSAnalyzerFile(QString filename, QVector<CANFrame>* frames)
 {
     QFile *inFile = new QFile(filename);
-    CANFrame thisFrame;
-    QString line;
-    int lineCounter = 0;
-    bool foundErrors = false;
-
     if (!inFile->open(QIODevice::ReadOnly))
```
Comment (KostyaSha) [inline framefileio.cpp:771]: firstly attempt to open file and only then create local vars

**Comment by KostyaSha** [comment]
Screen Shot 2020-10-07 at 02 59 50
It that's correct time, then i'm done. Though i didn't test with real adapters yet, but format looks good now.

**Comment by KostyaSha** [comment]
Screen Shot 2020-10-07 at 03 21 30
Aligned data

File: framefileio.cpp [hunk 3]
```diff
@@ -33,9 +34,10 @@ bool FrameFileIO::saveFrameFile(QString &fileName, const QVector<CANFrame>* fram
     filters.append(QString(tr("IXXAT MiniLog (*.csv *.CSV)")));
     filters.append(QString(tr("CAN-DO Log (*.can *.avc *.evc *.qcc *.CAN *.AVC *.EVC *.QCC)")));
     filters.append(QString(tr("Vehicle Spy (*.csv *.CSV)")));
-    filters.append(QString(tr("Candump/Kayak(*.log)")));
-    filters.append(QString(tr("Cabana Log(*.csv *.CSV)")));
+    filters.append(QString(tr("Candump/Kayak (*.log)")));
+    filters.append(QString(tr("Cabana Log (*.csv *.CSV)")));
```
Comment (KostyaSha) [inline framefileio.cpp:38]: on osx all listed files are in background color even if the match. Have no idea whats wrong

**Comment by KostyaSha** [comment]
@collin80 i'm done

**Pull request merged**